src/formats/onenote.ts

```diff
@@ -0,0 +1,489 @@
+const ATTACHMENT_EXTS = ['png', 'webp', 'jpg', 'jpeg', 'gif', 'bmp', 'svg', 'mpg', 'm4a', 'webm', 'wav', 'ogv', '3gp', 'mov', 'mp4', 'mkv', 'pdf'];
+
+export class OneNoteImporter extends FormatImporter {
+	useDefaultAttachmentFolder: boolean = true;
+	importIncompatibleAttachments: boolean;
+	contentArea: HTMLDivElement;
+	requestParams: RequestInit;
+	attachmentQueue: {name: string, url: string } [] = [];
+	selectedSections: OnenoteSection[] = [];
+
+	init() {
+		this.addOutputLocationSetting('OneNote');
+
+		this.showUI();
+	}
+
+	async showUI() {
+		new Setting(this.modal.contentEl)
+			.setName('Use the default attachment folder')
+			.setDesc('If disabled, attachments will be stored in the export folder in the OneNote Attachments folder.')
+			.addToggle((toggle) => toggle
+				.setValue(true)
+				.onChange((value) => (this.useDefaultAttachmentFolder = value))
+			);
+		new Setting(this.modal.contentEl)
+			.setName('Import incompatible attachments')
+			.setDesc('Imports incompatible attachments which cannot be embedded in Obsidian, such as .exe files.')
+			.addToggle((toggle) => toggle
+				.setValue(false)
+				.onChange((value) => (this.importIncompatibleAttachments = value))
+			);
+
+		// Create a wrapper for sign in related settings in order to hide them later
+		this.contentArea = this.modal.contentEl.createEl('div');
+
+		// Create an description with an link
+		let descriptionFragment = new DocumentFragment();
+		descriptionFragment.createSpan({ text: 'Go to ' });
+		descriptionFragment.createEl('a', {
+			text: 'microsoft.com/devicelogin',
+			href: 'https://microsoft.com/devicelogin',
+		});
+		descriptionFragment.createSpan({ text: ' and enter the code visible on the right side.' });
+	
+		const loginCode = await this.generateLoginCode();
+
+		// Add an separator
+		this.contentArea.createEl('span');
+		
+		new Setting(this.contentArea)
+			.setName('Sign in to your Microsoft Account')
+			.setDesc(descriptionFragment)
+			.addText((text) => text
+				.setValue(loginCode)
+				.setDisabled(true))
+			.addButton((button) => button
+				.setCta()
+				.setButtonText('Copy')
+				.onClick(() => navigator.clipboard.writeText(loginCode))
+			);
+
+		new Setting(this.contentArea)
+			.setName('Custom user access token')
+			.setDesc('If you are having troubles with the device code, use a custom access token from Microsoft Graph Explorer by going to the access token tab under the address bar.')
+			.addText((text) => text.setPlaceholder('Paste token here')
+				.onChange(async (e) => this.signIn(e)));
+	}
+
+	async generateLoginCode(): Promise<string> {
+		const requestBody = new URLSearchParams({
+			client_id: GRAPH_CLIENT_ID,
+			scope: GRAPH_SCOPES.join(' '),
+		});
+
+		// Using requestUrl in order to prevent CORS issues
+		const tokenResponse = await requestUrl(
+			`https://login.microsoftonline.com/common/oauth2/v2.0/devicecode?${requestBody.toString()}`
```

Comment:
I'm concerned about the security and possibility of phishing with this auth method. The client ID is embedded in the plugin and could be used outside of this plugin by anyone. From my research, this auth method is intended only for limited-input devices (e.g. TVs).

Please use the [auth code grant](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) authentication flow. Obsidian supports plugins [registering custom URIs](https://help.obsidian.md/Concepts/Obsidian+URI) which can be used for the auth redirect.

Reply:
See also https://docs.obsidian.md/Reference/TypeScript+API/Plugin/registerObsidianProtocolHandler